spark_log_parser/loaders/__init__.py

```diff
@@ -61,58 +71,110 @@ def __iter__(self):
+                self._buffer = self._buffer[self._buffer_position:]
+                self._buffer_position = 0
+
+                # iterate chunks until we've read at least `size` amount of data
+                for chunk in self._chunks:
+                    self._add_chunk_to_size(chunk)
+                    self._buffer += chunk
+                    if len(self._buffer) >= size:
+                        break
+
+                # Freeze this buffer for subsequent reads from it. We use a memoryview in order to avoid making copies
+                #  of the data for every single read issued.
+                self._buffer_memview = memoryview(self._buffer).cast("B").toreadonly()
+
+            # If we have enough data in the buffer to satisfy this read, return that from the buffer and adjust
+            #  our read position accordingly
+            if self._num_bytes_left_in_buffer() >= size:
+                start_pos = self._buffer_position
+                self._buffer_position += size
+                return self._buffer_memview[start_pos:self._buffer_position]
+            else:
+                # Otherwise, we don't have enough data, which means that this is the last read for this iterator.
+                #  So now just return a copy of the data remaining in the buffer and mark this as consumed to prevent
+                #  future reads
+                content = self._buffer_memview[self._buffer_position:]
+                self._consumed = True
+                return content
+
+    _LINE_CHUNK_SIZE = 1024 * 1024
 
     def iter_lines(self) -> Iterator[bytes]:
         """
         Returns a stream of lines from self._chunks.
         """
-        for chunk in self._chunks:
-            self._add_chunk_to_size(chunk)
-            if self._trailing_data:
-                chunk = self._trailing_data + chunk
-
-            lines = chunk.splitlines()
+        trailing_data = bytearray()
+        # We use a relatively large chunk_size here to minimize the number of times we have to enter this loop/copy data
+        while chunk := self.read(self._LINE_CHUNK_SIZE):
+            if trailing_data:
+                # Instead of appending via e.g. chunk = trailing_data + chunk, which causes a ton of data copying,
+                #  we can just write this chunk into our current line buffer. This is extremely helpful when we
+                #  encounter large files where all the data is actually on one line (i.e. a 500MB JSON file that was
+                #  written without any line breaks). This saves us 10s of seconds when processing these types of files
+                # This is safe to do here because trailing_data is never referenced outside of this function, and we
+                #  only ever return trailing_data when we are done reading in the whole file
+                trailing_data.extend(chunk)
+                chunk = trailing_data
+                trailing_data = bytearray()
```

Comment:
There is a big comment here about why I'm doing this, but also to demonstrate clearly the benefit,
here are two side-by-side `cprof` captures against a parsed eventlog that Insider sent over to us just last week. Just this change, re-using and calling `.extend` on our byte array, saved 26 _**seconds**_ of pure CPU time (old implementation on the left, new on the right) -
 
<img width="1392" alt="Screenshot 2023-01-24 at 1 07 50 PM" src="https://user-images.githubusercontent.com/117235708/214428182-6464213c-1351-437f-9d23-c23f54cf4ead.png">